test/unit/vm/disposable_test.dart

```diff
@@ -103,44 +124,112 @@ void main() {
       });
     });
 
+    void testManageMethod(
+        String methodName, callback(dynamic argument), dynamic argument,
+        {bool doesCallbackReturn: true}) {
+      if (doesCallbackReturn) {
+        test('should return the argument', () {
+          expect(callback(argument), same(argument));
+        });
+      }
+
+      test('should throw if called with a null argument', () {
+        expect(() => callback(null), throwsArgumentError);
+      });
+
+      test('should throw if object is disposing', () async {
+        thing.manageDisposer(() async {
+          expect(() => callback(argument), throwsStateError);
+        });
+        await thing.dispose();
+      });
+
+      test('should throw if object has been disposed', () async {
+        await thing.dispose();
+        expect(() => callback(argument), throwsStateError);
+      });
+    }
+
+    group('awaitBeforeDispose', () {
+      test('should wait for the future to complete before disposing', () async {
+        var completer = new Completer();
+        var awaitedFuture = thing.awaitBeforeDispose(completer.future);
+        var disposeFuture = thing.dispose().then((_) {
+          expect(thing.isDisposing, isFalse,
+              reason: 'isDisposing post-complete');
+          expect(thing.isDisposed, isTrue, reason: 'isDisposed post-complete');
+        });
+        await new Future(() {});
+        expect(thing.isDisposing, isTrue, reason: 'isDisposing pre-complete');
+        expect(thing.isDisposed, isFalse, reason: 'isDisposed pre-complete');
+        completer.complete();
+        // It's simpler to do this than ignore a bunch of lints.
+        await Future.wait([awaitedFuture, disposeFuture]);
+      });
+
+      testManageMethod('waitBeforeDispose',
+          (argument) => thing.awaitBeforeDispose(argument), new Future(() {}));
+    });
+
+    group('manageCompleter', () {
+      test('should complete with an error when parent is disposed', () {
+        var completer = new Completer<Null>();
+        completer.future.catchError(expectAsync1((exception) {
+          expect(exception, new isInstanceOf<ObjectDisposedException>());
+        }));
+        thing.manageCompleter(completer);
+        thing.dispose();
```

Comment:
Should `dispose` return any error it encountered, such as the one above? A consume disposing something may want to handle a dirty disposal.